go/tools/builders/nogo_main.go

```diff
@@ -202,6 +202,55 @@ func checkPackage(analyzers []*analysis.Analyzer, packagePath string, packageFil
 		act.pkg = pkg
 	}
 
+	// Process nolint directives similar to golangci-lint.
+	for _, f := range pkg.syntax {
+		// First, gather all comments and apply a range to the entire comment block.
+		// This will cover inline comments and allow us to detect comments above a
+		// node later.
+		for _, group := range f.Comments {
+			for _, comm := range group.List {
+				linters, ok := parseNolint(comm.Text)
+				if !ok {
+					continue
+				}
+				nl := &Range{
+					from: pkg.fset.Position(group.Pos()),
+					to:   pkg.fset.Position(group.End()).Line,
+				}
+				for analyzer, act := range actions {
+					if linters.Contains(analyzer.Name) {
+						act.nolint = append(act.nolint, nl)
+					}
+				}
+			}
+		}
+
+		// For each node in the ast, check if the previous line is covered by a
+		// range for the linter and expand it.
+		ast.Inspect(f, func(n ast.Node) bool {
+			if n == nil {
+				return true
+			}
+
+			nodeStart := pkg.fset.Position(n.Pos())
+			nodeEnd := pkg.fset.Position(n.End())
+			for _, act := range actions {
+				for _, rng := range act.nolint {
+					if rng.from.Filename != nodeStart.Filename {
+						continue
+					}
```

Comment:
In which scenarios does the File name differ here?

I thought we are still in the same `for _, f := range pkg.syntax {` loop where `f == file`? 🤔
So these should all have the same file name no?

Reply:
The ranges are built for the entire package across multiple files. I didn't want a range for one file to apply to an error in a different file with a similar line number.